src/main/scala/firrtl/graph/DiGraph.scala

```diff
@@ -34,7 +34,11 @@ object DiGraph {
 }
 
 /** Represents common behavior of all directed graphs */
-class DiGraph[T] private[graph] (private[graph] val edges: LinkedHashMap[T, LinkedHashSet[T]]) {
+trait DiGraphLike[T] {
+  private[graph] val edges: LinkedHashMap[T, LinkedHashSet[T]]
+
+  private[graph] val prev: LinkedHashMap[T, T]
```

Comment:
This isn't used in anything operating on a `DiGraphLike`, so maybe it should only be in a separate trait extended by the derived classes that use it? As far as I can tell, it looks like it could even be private to the derived classes.

Reply:
I made `DiGraph` to public and abstract to solve this issue, is this a good way to do so?

Reply:
I guess I mean why does `DiGraph` have `prev` if it isn't used?

Reply:
I think there still needs some codes clean up after I fixes some tests.